hypothesis-python/tests/ghostwriter/test_ghostwriter_cli.py

```diff
@@ -116,6 +121,162 @@ def test_can_import_from_scripts_in_working_dir(tmpdir):
+        ),
+        (
+            "func_sorter",
+            "XX",
+            "Found the 'mycode' module and 'func_sorter' attribute, but it doesn't have a 'XX' attribute.",
+        ),
+    ],
+)
+def test_error_import_from_class(tmpdir, classname, funcname, err_msg):
+    (tmpdir / "mycode.py").write(CLASS_CODE_TO_TEST)
+    result = subprocess.run(
+        f"hypothesis write mycode.{classname}.{funcname}",
+        capture_output=True,
+        shell=True,
+        text=True,
+        cwd=tmpdir,
+    )
+    assert result.returncode == 2
+    assert "Error: " + err_msg in result.stderr
+
+
+def test_magic_discovery_from_module(tmpdir):
+    (tmpdir / "mycode.py").write(CLASS_CODE_TO_TEST)
+    result = subprocess.run(
+        f"hypothesis write mycode",
+        capture_output=True,
+        shell=True,
+        text=True,
+        cwd=tmpdir,
+    )
+    assert result.returncode == 0
+    assert "func_sorter" in result.stdout
+    assert "S.static_sorter" in result.stdout
+    assert "S.class_sorter" in result.stdout
+
+
+ROUNDTRIP_CODE_TO_TEST = """
+from typing import Union
+import json
+
+def to_json(json: Union[dict,list]) -> str:
+    return json.dumps(json)
+
+def from_json(json: str) -> Union[dict,list]:
+    return json.loads(json)
+
+class S:
+
+    @staticmethod
+    def to_json(json: Union[dict,list]) -> str:
+        return json.dumps(json)
+
+    @staticmethod
+    def from_json(json: str) -> Union[dict,list]:
+        return json.loads(json)
+
+class A:
+
+    @classmethod
+    def to_json(cls, json: Union[dict,list]) -> str:
+        return json.dumps(json)
+
+    @classmethod
+    def from_json(cls, json: str) -> Union[dict,list]:
+        return json.loads(json)
+"""
+
+
+def test_roundtrip_correct_pairs(tmpdir):
+    (tmpdir / "mycode.py").write(ROUNDTRIP_CODE_TO_TEST)
+    result = subprocess.run(
+        f"hypothesis write mycode",
+        capture_output=True,
+        shell=True,
+        text=True,
+        cwd=tmpdir,
+    )
+    assert result.returncode == 0
+    # correct paris
+    assert (
+        """value0 = mycode.S.to_json(json=json)
+    value1 = mycode.S.from_json(json=value0)"""
+        in result.stdout
+    )
+    assert (
+        """value0 = mycode.A.to_json(json=json)
+    value1 = mycode.A.from_json(json=value0)"""
+        in result.stdout
+    )
+    assert (
+        """value0 = mycode.to_json(json=json)
+    value1 = mycode.from_json(json=value0)"""
+        in result.stdout
+    )
+    # incorrect paris
+    assert (
+        """value0 = mycode.to_json(json=json)
+    value1 = mycode.S.from_json(json=value0)"""
+        not in result.stdout
+    )
+    assert (
+        """value0 = mycode.S.to_json(json=json)
+    value1 = mycode.from_json(json=value0)"""
+        not in result.stdout
+    )
+    assert (
+        """value0 = mycode.S.to_json(json=json)
+    value1 = mycode.A.from_json(json=value0)"""
+        not in result.stdout
+    )
```

Comment:
Can we make each of these sets of checks into a loop, using just the short parts to format into a common string?

Reply:
yup